master/internal/api_experiment.go

```diff
@@ -708,3 +710,172 @@ func (a *apiServer) TrialsSnapshot(req *apiv1.TrialsSnapshotRequest,
 		}
 	}
 }
+
+func (a *apiServer) topTrials(experimentID int, maxTrials int, s model.SearcherConfig) (
+	trials []int32, err error) {
+	type Ranking string
+	const (
+		ByMetricOfInterest Ranking = "ByMetricOfInterest"
+		ByTrainingLength   Ranking = "ByTrainingLength"
+	)
+	var ranking Ranking
+
+	switch {
+	case s.RandomConfig != nil:
+		ranking = ByMetricOfInterest
+	case s.GridConfig != nil:
+		ranking = ByMetricOfInterest
+	case s.SyncHalvingConfig != nil:
+		ranking = ByTrainingLength
+	case s.AdaptiveConfig != nil:
+		ranking = ByTrainingLength
+	case s.AdaptiveSimpleConfig != nil:
+		ranking = ByTrainingLength
+	case s.AsyncHalvingConfig != nil:
+		ranking = ByTrainingLength
+	case s.AdaptiveASHAConfig != nil:
+		ranking = ByTrainingLength
+	case s.SingleConfig != nil:
+		return nil, errors.New("single-trial experiments are not supported for trial sampling")
+	case s.PBTConfig != nil:
+		return nil, errors.New("population-based training not supported for trial sampling")
+	default:
+		return nil, errors.New("unable to detect a searcher algorithm for trial sampling")
+	}
+	switch ranking {
+	case ByMetricOfInterest:
+		return a.m.db.TopTrialsByMetric(experimentID, maxTrials, s.Metric, s.SmallerIsBetter)
+	case ByTrainingLength:
+		return a.m.db.TopTrialsByTrainingLength(experimentID, maxTrials, s.Metric, s.SmallerIsBetter)
+	default:
+		panic("Invalid state in trial sampling")
+	}
+}
+
+func (a *apiServer) TrialsSample(req *apiv1.TrialsSampleRequest,
+	resp apiv1.Determined_TrialsSampleServer) error {
+	experimentID := int(req.ExperimentId)
+	maxTrials := int(req.MaxTrials)
+	if maxTrials == 0 {
+		maxTrials = 25
+	}
+	maxDatapoints := int(req.MaxDatapoints)
+	if maxDatapoints == 0 {
+		maxDatapoints = 1000
+	}
+	startBatches := int(req.StartBatches)
+	endBatches := int(req.EndBatches)
+	if endBatches <= 0 {
+		endBatches = math.MaxInt32
+	}
+
+	metricName := req.MetricName
+	metricType := req.MetricType
+	if metricType == apiv1.MetricType_METRIC_TYPE_UNSPECIFIED {
+		return errors.New("must specify a metric type")
+	}
+	if metricName == "" {
+		return errors.New("must provide metric name")
+	}
+
+	config, err := a.m.db.ExperimentConfig(experimentID)
+	if err != nil {
+		return errors.Wrapf(err, "error fetching experiment config from database")
+	}
+	searcherConfig := config.Searcher
+
+	trialCursors := make(map[int32]time.Time)
+	currentTrials := make(map[int32]bool)
+	var zeroTime time.Time
+	for {
+		var response apiv1.TrialsSampleResponse
+		var promotedTrials []int32
+		var demotedTrials []int32
+		var trials []*apiv1.TrialsSampleResponse_Trial
+
+		seenThisRound := make(map[int32]bool)
+
+		var metricSeries []lttb.Point
+		var endTime time.Time
+		trialIDs, err := a.topTrials(experimentID, maxTrials, searcherConfig)
+		if err != nil {
+			return errors.Wrapf(err, "error determining top trials")
+		}
+		for _, trialID := range trialIDs {
+			var trial apiv1.TrialsSampleResponse_Trial
```

Comment:
non-blocking: what do you think about pulling this inner blocking out into like `trialSample(trialID int...) apiv1.TrialsSampleResponse_Trial` (plus everything else it needs too?). It's a lot to grok in one sitting for slow readers like me lol.

Reply:
Good one. So all the logic of tracking which trials were being promoted / demoted etc, is now in the original function, and the logic that pertains to a single trial is factored out into a new function.